diff -Naur dropbear-2014.63/cli-authpasswd.c dropbear-2014.63-ltrx/cli-authpasswd.c
--- dropbear-2014.63/cli-authpasswd.c	2014-02-19 06:05:24.000000000 -0800
+++ dropbear-2014.63-ltrx/cli-authpasswd.c	2016-01-13 23:18:32.213466985 -0800
@@ -134,6 +134,11 @@
 		}
 	} else
 #endif
+     if(cli_opts.optPassword && *cli_opts.optPassword)
+     {
+         password = strdup(cli_opts.optPassword);
+     }
+     else
 	{
 		password = getpass_or_cancel(prompt);
 	}
diff -Naur dropbear-2014.63/cli-runopts.c dropbear-2014.63-ltrx/cli-runopts.c
--- dropbear-2014.63/cli-runopts.c	2016-01-13 23:13:28.777458337 -0800
+++ dropbear-2014.63-ltrx/cli-runopts.c	2016-01-13 23:20:55.721469633 -0800
@@ -126,6 +126,7 @@
 	cli_opts.remotehost = NULL;
 	cli_opts.remoteport = NULL;
 	cli_opts.username = NULL;
+	cli_opts.optPassword = NULL;
 	cli_opts.cmd = NULL;
 	cli_opts.no_cmd = 0;
 	cli_opts.backgrounded = 0;
@@ -280,6 +281,10 @@
 				case 'l':
 					next = &cli_opts.username;
 					break;
+                 case 'P':
+                     next = &cli_opts.optPassword;
+                     break;
+
 				case 'h':
 					printhelp();
 					exit(EXIT_SUCCESS);
diff -Naur dropbear-2014.63/runopts.h dropbear-2014.63-ltrx/runopts.h
--- dropbear-2014.63/runopts.h	2016-01-13 23:13:28.777458337 -0800
+++ dropbear-2014.63-ltrx/runopts.h	2016-01-13 23:21:32.805470797 -0800
@@ -125,6 +125,7 @@
 
 	char *own_user;
 	char *username;
+	char *optPassword;
 
 	char *cmd;
 	int wantpty;
