diff -Naur busybox-1.22.1/include/applets.src.h busybox-1.22.1_ltrx/include/applets.src.h
--- busybox-1.22.1/include/applets.src.h	2016-01-22 01:23:42.727356613 -0800
+++ busybox-1.22.1_ltrx/include/applets.src.h	2016-01-22 01:41:42.751383600 -0800
@@ -109,6 +109,7 @@
 IF_CRONTAB(APPLET(crontab, BB_DIR_USR_BIN, BB_SUID_REQUIRE))
 IF_CRYPTPW(APPLET(cryptpw, BB_DIR_USR_BIN, BB_SUID_DROP))
 IF_CUT(APPLET_NOEXEC(cut, cut, BB_DIR_USR_BIN, BB_SUID_DROP, cut))
+IF_DATE(APPLET(date, BB_DIR_BIN, BB_SUID_REQUIRE))
 IF_DC(APPLET(dc, BB_DIR_USR_BIN, BB_SUID_DROP))
 IF_DD(APPLET_NOEXEC(dd, dd, BB_DIR_BIN, BB_SUID_DROP, dd))
 IF_DEALLOCVT(APPLET(deallocvt, BB_DIR_USR_BIN, BB_SUID_DROP))
